Fixes for the x64 architecture

eloraiby opened the pull request:
Hold the fixes for the x64 architecture.

eloraiby committed:
Merge from upstream

eloraiby committed:
Fix few warnings from the static analyzer for the x64 architecture

eloraiby committed:
Add PushSizeT and related macros/functions
tOps now holds size_t* ops instead of U32*
Silence warnings related to unreferenced labels in JIT_Execute.c

ncave commented on src/DNA/native/src/JIT_Execute.c:
```diff
@@ -694,31 +703,31 @@ U32 JIT_Execute(tThread *pThread, U32 numInst) {
-JIT_CONV_I64_U64_end:
-JIT_CONV_U64_I64_end:
-JIT_GoNext_start:
+//JIT_NOP_end:
```
@eloraiby Do you need to comment the ends, I was trying to keep the threaded code intact in case it's faster than the switch on some platforms? Is there another way to ignore the unused label warnings?

ncave replied:
Unless of course you just want to get rid of it.

eloraiby replied:
@ncave I can do it via a macro :), probably that's better for now. I would hold before cleaning up the code entirely, at least until we have a working x64 version. I haven't finished merging the upstream yet.

eloraiby committed:
Remove the m32 compilation

eloraiby committed:
Merge branch 'master' of https://github.com/ncave/dna-redux

eloraiby committed:
Remove trailing #endif after merge

eloraiby committed:
Static Analysis: Fix bug in InternalCall.c - unhandled case where curType, curNameSpace could be null

eloraiby committed:
Static Analysis - Bug fix :
- LoadDebugFile : unhandled case when pPrevious == NULL
- LoadPEFile    : unhandled case when pData == NULL

eloraiby commented:
@ncave still a long way to go before making it run on x64, but few bug fix are in thanks to static analysis (while one should never happen, I kept it just to be safe)

eloraiby commented:
@ncave A complicated bug in the garbage collector pointed by the static analyzer. Need your opinion before any attempt to fix:
```
dna-redux/src/DNA/native/src/Heap.c:398:19: warning: Use of memory after it is freed [clang-analyzer-unix.Malloc]
                pHeapTreeRoot = TreeRemove(pHeapTreeRoot, pThis);
                                ^
dna-redux/src/DNA/native/src/Heap.c:538:10: note: Calling 'Heap_Alloc'
        clone = Heap_Alloc(pObj->pTypeDef, size);
                ^
dna-redux/src/DNA/native/src/Heap.c:472:2: note: Taking true branch
        if (trackHeapSize >= heapSizeMax) {
        ^
dna-redux/src/DNA/native/src/Heap.c:473:3: note: Calling 'GarbageCollect'
                GarbageCollect();
                ^
dna-redux/src/DNA/native/src/Heap.c:420:2: note: Calling 'GC_Sweep'
        GC_Sweep();
        ^
dna-redux/src/DNA/native/src/Heap.c:344:2: note: Loop condition is true.  Entering loop body
        while (top != 0) {
        ^
dna-redux/src/DNA/native/src/Heap.c:348:7: note: Assuming the condition is false
                if (pNode->marked) {
                    ^
dna-redux/src/DNA/native/src/Heap.c:348:3: note: Taking false branch
                if (pNode->marked) {
                ^
dna-redux/src/DNA/native/src/Heap.c:360:8: note: Assuming the condition is false
                        if (pNode->needToFinalize) {
                            ^
dna-redux/src/DNA/native/src/Heap.c:360:4: note: Taking false branch
                        if (pNode->needToFinalize) {
                        ^
dna-redux/src/DNA/native/src/Heap.c:375:9: note: Assuming the condition is false
                                if (pNode->pSync != NULL) {
                                    ^
dna-redux/src/DNA/native/src/Heap.c:375:5: note: Taking false branch
                                if (pNode->pSync != NULL) {
                                ^
dna-redux/src/DNA/native/src/Heap.c:386:7: note: Assuming the condition is false
                if (pNode->pLink[1] != nil) {
                    ^
dna-redux/src/DNA/native/src/Heap.c:386:3: note: Taking false branch
                if (pNode->pLink[1] != nil) {
                ^
dna-redux/src/DNA/native/src/Heap.c:389:7: note: Assuming the condition is true
                if (pNode->pLink[0] != nil) {
                    ^
dna-redux/src/DNA/native/src/Heap.c:389:3: note: Taking true branch
                if (pNode->pLink[0] != nil) {
                ^
dna-redux/src/DNA/native/src/Heap.c:344:2: note: Loop condition is true.  Entering loop body
        while (top != 0) {
        ^
dna-redux/src/DNA/native/src/Heap.c:348:7: note: Assuming the condition is false
                if (pNode->marked) {
                    ^
dna-redux/src/DNA/native/src/Heap.c:348:3: note: Taking false branch
                if (pNode->marked) {
                ^
dna-redux/src/DNA/native/src/Heap.c:360:8: note: Assuming the condition is false
                        if (pNode->needToFinalize) {
                            ^
dna-redux/src/DNA/native/src/Heap.c:360:4: note: Taking false branch
                        if (pNode->needToFinalize) {
                        ^
dna-redux/src/DNA/native/src/Heap.c:375:9: note: Assuming the condition is false
                                if (pNode->pSync != NULL) {
                                    ^
dna-redux/src/DNA/native/src/Heap.c:375:5: note: Taking false branch
                                if (pNode->pSync != NULL) {
                                ^
dna-redux/src/DNA/native/src/Heap.c:386:7: note: Assuming the condition is false
                if (pNode->pLink[1] != nil) {
                    ^
dna-redux/src/DNA/native/src/Heap.c:386:3: note: Taking false branch
                if (pNode->pLink[1] != nil) {
                ^
dna-redux/src/DNA/native/src/Heap.c:389:7: note: Assuming the condition is false
                if (pNode->pLink[0] != nil) {
                    ^
dna-redux/src/DNA/native/src/Heap.c:389:3: note: Taking false branch
                if (pNode->pLink[0] != nil) {
                ^
dna-redux/src/DNA/native/src/Heap.c:344:2: note: Loop condition is false. Execution continues on line 395
        while (top != 0) {
        ^
dna-redux/src/DNA/native/src/Heap.c:395:2: note: Loop condition is true.  Entering loop body
        while (pToDelete != NULL) {
        ^
dna-redux/src/DNA/native/src/Heap.c:398:19: note: Calling 'TreeRemove'
                pHeapTreeRoot = TreeRemove(pHeapTreeRoot, pThis);
                                ^
dna-redux/src/DNA/native/src/Heap.c:222:6: note: Assuming 'pRoot' is equal to 'nil'
        if (pRoot != nil) {
            ^
dna-redux/src/DNA/native/src/Heap.c:222:2: note: Taking false branch
        if (pRoot != nil) {
        ^
dna-redux/src/DNA/native/src/Heap.c:398:19: note: Returning from 'TreeRemove'
                pHeapTreeRoot = TreeRemove(pHeapTreeRoot, pThis);
                                ^
dna-redux/src/DNA/native/src/Heap.c:395:2: note: Loop condition is true.  Entering loop body
        while (pToDelete != NULL) {
        ^
dna-redux/src/DNA/native/src/Heap.c:401:3: note: Memory is released
                free(pThis);
                ^
dna-redux/src/DNA/native/src/Heap.c:395:9: note: Assuming 'pToDelete' is not equal to NULL
        while (pToDelete != NULL) {
               ^
dna-redux/src/DNA/native/src/Heap.c:395:2: note: Loop condition is true.  Entering loop body
        while (pToDelete != NULL) {
        ^
dna-redux/src/DNA/native/src/Heap.c:398:19: note: Use of memory after it is freed
                pHeapTreeRoot = TreeRemove(pHeapTreeRoot, pThis);
```

eloraiby commented:
btw, the following files are completed with static analysis and fixing (just warning related to the use of unsafe str* functions - c99):
```
    src/CLIFile.c
    src/Delegate.c
    src/dna.c
    src/Finalizer.c
    src/Generics.c
    src/InternalCall.c
```

ncave commented:
@eloraiby Awesome, thanks. Yes, this is quite complicated GC behavior, is it possible that it's a false positive? I'm only asking because I've tried static analysis before with the MS VS analyzer and it was tripping on most delayed allocations, so I kind of ignored those.

Your analysis looks way more detailed, so perhaps it's a real issue. Even if so, I'm not sure this is a high priority, so it's up to you if you want to spend time fixing it right now (unless it impedes your progress), or just open a new issue with the description, so we don't lose track of it.

Let me know when you want me to merge this PR.
As I'm sure you already know, you can always run [without any GC](https://github.com/ncave/dna-redux/blob/master/src/DNA/native/src/Config.h#L81) if you need maximum speed, or want to check if some issue is not GC related.

eloraiby commented:
@ncave please merge this one as to no loose progress. Will create another one for the beefy JIT.c JIT_Execute.c fixes.

eloraiby commented:
@ncave on a side note, I am using `clang-tidy` for static analysis. Quite good actually :)

ncave commented on src/DNA/native/src/JIT.c:
```diff
@@ -132,8 +133,19 @@ static void PushU32_(tOps *pOps, U32 v, I32 opSequencePoint) {
 		//dprintfn("a.pOps->p = 0x%08x size=%d", pOps->p, pOps->capacity * sizeof(U32));
-		pOps->p = realloc(pOps->p, pOps->capacity * sizeof(U32));
-		pOps->pSequencePoints = realloc(pOps->pSequencePoints, pOps->capacity * sizeof(U32));
+		pOps->p = realloc(pOps->p, pOps->capacity * sizeof(size_t));
```
Shouldn't this be U32 (here and next line) instead of size_t, since you introduced PushSizeT_ below?

ncave replied:
Nevermind, I see you changed the storage.
